src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -602,7 +602,16 @@ private StructImpl completeDownload(PendingDownload pendingDownload)
           "Downloads a file to the output path for the provided url and returns a struct"
               + " containing <code>success</code>, a flag which is <code>true</code> if the"
               + " download completed successfully, and if successful, a hash of the file"
-              + " with the fields <code>sha256</code> and <code>integrity</code>.",
+              + " with the fields <code>sha256</code> and <code>integrity</code>."
+              + "Setting an explicit <code>canonical_id</code> is highly recommended. e.g."
```

Comment:
I find this recommendation a bit too broad: If this were the case, wouldn't we always want to use canonical ID? Instead, I think that we should recommend it in the case where the hash is user-specified rather than generated automatically.

Reply:
I've amended the recommendation to be specific to when `sha256` or `integrity` are user specified.